Build errors on iOS and read errors on JVM

I have extended the previous test project which I sent you to also test resources with blanks in their names and I also test whether I can actually read the resource bytes and not just whether the resource exists. Doing that I now get two more failures. The result for all platforms now looks like this.

```
Target           , withBlank, withoutBlank
iosSimulatorArm64, FAIL1    , FAIL1
jvm              , FAIL2    , OK
js               , OK       , OK
wasmJs           , OK       , OK
Android (local)  , OK       , OK

FAIL1:
======
A problem was found with the configuration of task ':composeApp:iosSimulatorArm64DebugTestCopyResources' (type 'Copy').
  - Gradle detected a problem with the following location: '/Volumes/IP11ext/Downloads/tmp/AndroidResourcesBug/composeApp/build/generated/compose/resourceGenerator/assembledResources/iosSimulatorArm64Test'.

    Reason: Task ':composeApp:iosSimulatorArm64DebugTestCopyResources' uses this output of task ':composeApp:assembleIosSimulatorArm64TestResources' without declaring an explicit or implicit dependency. This can lead to incorrect results being produced, depending on what order the tasks are executed.

    Possible solutions:
      1. Declare task ':composeApp:assembleIosSimulatorArm64TestResources' as an input of ':composeApp:iosSimulatorArm64DebugTestCopyResources'.
      2. Declare an explicit dependency on ':composeApp:assembleIosSimulatorArm64TestResources' from ':composeApp:iosSimulatorArm64DebugTestCopyResources' using Task#dependsOn.
      3. Declare an explicit dependency on ':composeApp:assembleIosSimulatorArm64TestResources' from ':composeApp:iosSimulatorArm64DebugTestCopyResources' using Task#mustRunAfter.

    For more information, please refer to https://docs.gradle.org/9.3.0/userguide/validation_problems.html#implicit_dependency in the Gradle documentation.

* Try:
> Run with --scan to get full insights from a Build Scan (powered by Develocity).

Deprecated Gradle features were used in this build, making it incompatible with Gradle 10.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

For more on this, please refer to https://docs.gradle.org/9.3.0/userguide/command_line_interface.html#sec:command_line_warnings in the Gradle documentation.

FAIL2:
======
Caused by: java.io.FileNotFoundException: /Volumes/IP11ext/Downloads/tmp/AndroidResourcesBug/composeApp/build/processedResources/jvm/test/pdf/AD-2.LFGA%20page1.pdf (No such file or directory)
This happens when actually reading the resource.

```

[AndroidResourcesBug2.zip](https://github.com/user-attachments/files/24980404/AndroidResourcesBug2.zip)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -4,6 +4,10 @@ Notable changes are documented in this file, whose format follows [Keep a Change
 
 ## [Unreleased]
 
+### Fixed
+
+- Fix JVM resource loading when filenames contain spaces or other URL-special characters. ([#272](https://github.com/goncalossilva/kotlinx-resources/issues/272))
+
 ## [0.14.4] - 2026-01-28
 
 ### Fixed

--- a/resources-library/src/jvmMain/kotlin/Resource.kt
+++ b/resources-library/src/jvmMain/kotlin/Resource.kt
@@ -1,25 +1,25 @@
 package com.goncalossilva.resources
 
-import java.io.File
+import java.io.InputStream
 
 public actual class Resource actual constructor(private val path: String) {
 
-    private val resourcePath: String?
-        get() = Resource::class.java.classLoader.getResource(path)?.path
+    private val resource
+        get() = Resource::class.java.classLoader.getResource(path)
 
-    private val resourceFile: File
-        get() = File(resourcePath)
+    private val resourceStream: InputStream
+        get() = Resource::class.java.classLoader.getResourceAsStream(path)!!
 
-    public actual fun exists(): Boolean = resourcePath != null
+    public actual fun exists(): Boolean = resource != null
 
     public actual fun readText(charset: Charset): String = runCatching {
-        resourceFile.readText(charset.toJavaCharset())
+        resourceStream.use { it.readBytes().toString(charset.toJavaCharset()) }
     }.getOrElse { cause ->
         throw ResourceReadException("$path: No such file or directory", cause)
     }
 
     public actual fun readBytes(): ByteArray = runCatching {
-        resourceFile.readBytes()
+        resourceStream.use { it.readBytes() }
     }.getOrElse { cause ->
         throw ResourceReadException("$path: No such file or directory", cause)
     }

--- a/resources-test/src/commonTest/kotlin/ResourceTest.kt
+++ b/resources-test/src/commonTest/kotlin/ResourceTest.kt
@@ -128,6 +128,11 @@ class ResourceTest {
         assertEquals("H\u00E9llo", Resource("charset-utf8.txt").readText())
     }
 
+    @Test
+    fun readsResourceWithSpaces() {
+        assertEquals("hello", Resource("file with spaces.txt").readText().trim())
+    }
+
     companion object {
         const val JSON: String = "{}\n"
         val GZIP: ByteArray = byteArrayOf(

--- a/resources-test/src/jsTest/kotlin/JsResourceTest.kt
+++ b/resources-test/src/jsTest/kotlin/JsResourceTest.kt
@@ -12,8 +12,4 @@ class JsResourceTest {
         assertEquals("js", Resource("platform_resource.txt").readText())
     }
 
-    @Test
-    fun readsResourceWithSpaces() {
-        assertEquals("hello", Resource("file with spaces.txt").readText().trim())
-    }
 }

--- a/resources-test/src/jvmTest/kotlin/JvmResourceTest.kt
+++ b/resources-test/src/jvmTest/kotlin/JvmResourceTest.kt
@@ -11,4 +11,5 @@ class JvmResourceTest {
         assertTrue(Resource("platform_resource.txt").exists())
         assertEquals("jvm", Resource("platform_resource.txt").readText())
     }
+
 }

--- a/resources-test/src/wasmJsTest/kotlin/WasmJsResourceTest.kt
+++ b/resources-test/src/wasmJsTest/kotlin/WasmJsResourceTest.kt
@@ -12,8 +12,4 @@ class WasmJsResourceTest {
         assertEquals("wasmjs", Resource("platform_resource.txt").readText())
     }
 
-    @Test
-    fun readsResourceWithSpaces() {
-        assertEquals("hello", Resource("file with spaces.txt").readText().trim())
-    }
 }

